### PR TITLE
Adding new test for '--zero-metadata' command

### PR DESCRIPTION
**Pull request by karolinavelkaja**
Signed-off-by: Karolina Rogowska <karolina.rogowska@intel.com>

### CODE REVIEW

File: test/functional/tests/cli/test_zero_metadata_command.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+    with TestRun.step("Try to zero metadata on system disk."):
+        os_disks = get_system_disks()
+        for os_disk in os_disks:
+            output = TestRun.executor.run(cli.zero_metadata_cmd(str(os_disk)))
+            if output.exit_code == 0:
+                TestRun.LOGGER.error("Zeroing metadata should fail!")
+            else:
+                cli_messages.check_stderr_msg(output, cli_messages.error_handling)
+                TestRun.LOGGER.info("Cannot zero metadata on system disk as expected.")
```
Comment (Ostrokrzew) [inline test/functional/tests/cli/test_zero_metadata_command.py:56]: This conditional statement doesn't look fine to me. The default message shall be error message and I think You should place here try-except statement instead.

File: test/functional/api/cas/cli_messages.py [hunk 2]
```diff
@@ -83,6 +82,38 @@
...
 
+unavailable_device = [
+    r"Error while opening \'\S+\'exclusively. This can be due to\n"
+    r"cache instance running on this device. In such case please stop the cache and try again\."
+]
+
+unavailable_device = [
+    r"Error while opening \'\S+\'exclusively\. This can be due to\n"
+    r"cache instance running on this device\. In such case please stop the cache and try again\."
+]
```
Comment (Ostrokrzew) [inline test/functional/api/cas/cli_messages.py:93]: redundant - delete the first message

File: test/functional/tests/cli/test_zero_metadata_command.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+    with TestRun.step("Create filesystem on core device."):
+        core.create_filesystem(filesystem)
+
+    with TestRun.step("Zero metadata on core device and check if filesystem still exists"):
+        output = TestRun.executor.run(cli.zero_metadata_cmd(str(core.path)))
+        if output.exit_code == 0:
+            TestRun.LOGGER.error("Zeroing metadata should fail!")
+        else:
+            cli_messages.check_stderr_msg(output, cli_messages.no_cas_metadata)
+            TestRun.LOGGER.info("Cannot zero metadata as expected.")
```
Comment (Ostrokrzew) [inline test/functional/tests/cli/test_zero_metadata_command.py:104]: I have similar concern here as well - wouldn't try-except be better here?

File: test/functional/tests/cli/test_zero_metadata_command.py [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+    with TestRun.step("Mount core."):
+        core.mount(mount_point)
+
+    with TestRun.step("Zero metadata on mounted core device and validate result"):
+        output = TestRun.executor.run(cli.zero_metadata_cmd(str(core.path)))
+        if output.exit_code == 0:
+            TestRun.LOGGER.error("Zeroing metadata should fail!")
+        else:
+            cli_messages.check_stderr_msg(output, cli_messages.unavailable_device)
+            TestRun.LOGGER.info("Cannot zero metadata as expected.")
```
Comment (Ostrokrzew) [inline test/functional/tests/cli/test_zero_metadata_command.py:121]: as above

File: test/functional/tests/cli/test_zero_metadata_command.py [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+        except CmdException:
+            TestRun.LOGGER.info("Start cache failed as expected.")
+
+    with TestRun.step("Zero metadata on CAS device without force"):
+        output = TestRun.executor.run(cli.zero_metadata_cmd(str(cache_disk.path)))
+        if output.exit_code == 0:
+            TestRun.LOGGER.error("Zeroing metadata without force should fail!")
+        else:
+            cli_messages.check_stderr_msg(output, cli_messages.cache_dirty_data)
+            TestRun.LOGGER.info("Cannot zero metadata (without force) as expected.")
```
Comment (Ostrokrzew) [inline test/functional/tests/cli/test_zero_metadata_command.py:172]: as above

File: test/functional/tests/cli/test_zero_metadata_command.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+                TestRun.LOGGER.error("Start cache failed.")
+
+
+@pytest.mark.require_disk("cache", DiskTypeSet([DiskType.optane, DiskType.nand]))
+@pytest.mark.require_disk("core", DiskTypeLowerThan("cache"))
+def test_zero_metadata_dirty_shutdown():
+    """
+        title: Test for '--zero-metadata' and dirty shutdown scenario.
+        description: |
+          Test for '--zero-metadata' with&without 'force' option on cache which had been dirty shut
```
Comment (Ostrokrzew) [inline test/functional/tests/cli/test_zero_metadata_command.py:195]: missing spaces

File: test/functional/tests/cli/test_zero_metadata_command.py [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+        except CmdException:
+            TestRun.LOGGER.info("Start cache failed as expected.")
+
+    with TestRun.step("Zero metadata on CAS device without force"):
+        output = TestRun.executor.run(cli.zero_metadata_cmd(str(cache_dev.path)))
+        if output.exit_code == 0:
+            TestRun.LOGGER.error("Zeroing metadata without force should fail!")
+        else:
+            cli_messages.check_stderr_msg(output, cli_messages.cache_dirty_shutdown)
+            TestRun.LOGGER.info("Cannot zero metadata (without force) as expected.")
```
Comment (Ostrokrzew) [inline test/functional/tests/cli/test_zero_metadata_command.py:240]: the same suggestion

File: test/functional/tests/cli/test_zero_metadata_command.py [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+
+    with TestRun.step("Zero metadata on cache device with force"):
+        output = TestRun.executor.run(cli.zero_metadata_cmd(str(cache_dev.path), force=True))
+        if output.exit_code != 0:
+            TestRun.LOGGER.error("Zero metadata with force should works for cache device!")
+        else:
+            TestRun.LOGGER.info("Zeroing metadata with force ran successfully!")
```
Comment (Ostrokrzew) [inline test/functional/tests/cli/test_zero_metadata_command.py:247]: like above - with `stdout`

File: test/functional/tests/cli/test_zero_metadata_command.py [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+        try:
+            cache = casadm.start_cache(cache_dev, CacheMode.WT)
+            TestRun.LOGGER.info("Cache started successfully.")
+        except CmdException:
+            TestRun.LOGGER.error("Start cache failed.")
+
+
+def prepare_devices():
+    cache_disk = TestRun.disks['cache']
+    cache_disk.create_partitions([Size(5, Unit.GibiByte)])
```
Comment (Ostrokrzew) [inline test/functional/tests/cli/test_zero_metadata_command.py:259]: Cache's capacity can be lower than core's.

**Review by Ostrokrzew — CHANGES_REQUESTED** [review]

File: test/functional/tests/cli/test_zero_metadata_command.py (hunk 1 above)
Comment (karolinavelkaja) [reply test/functional/tests/cli/test_zero_metadata_command.py:56]: Good point to changing order of messages (else for error), thanks!
But for the try-except I think it is not suitable here (or I should perform commands different for that - is that what you meant? Using "TestRun.executor.run()" ends well (we are getting output from the command which were perform and not the 'CmdException' for catch) and the output consists error message - which we want to check.

Comment (Ostrokrzew) [reply test/functional/tests/cli/test_zero_metadata_command.py:56]: I think I meant something like this:
``` python
output = TestRun.executor.run(cli.zero_metadata_cmd(str(os_disk)))
try:
    cli_messages.check_stderr_msg(output, cli_messages.error_handling)
    TestRun.LOGGER.info("Cannot zero metadata on system disk as expected.")
except CmdException:    # not sure about the name of this custom exception
    TestRun.LOGGER.error("Zeroing metadata should fail!")
```
because `cli_messages.check_stderr_msg` can raise error.

Comment (karolinavelkaja) [reply test/functional/tests/cli/test_zero_metadata_command.py:56]: I've checked this, still the same - for failing case command performed correctly with error information logged "ERROR: Message is incorrect,..." and is not catched as CmdException.

File: test/functional/api/cas/cli_messages.py [hunk 10]
```diff
@@ -83,6 +82,33 @@
...
+
+no_cas_metadata = [
+    r"Device \'\S+\' does not contain OpenCAS's metadata\."
+]
+
+cache_dirty_data = [
+    r"Cache instance contains dirty data\. Clearing metadata will result in loss of dirty data\.\n"
+    r"Please load cache instance and flush dirty data in order to preserve them on the core "
+    r"device\.\n"
+    r"Alternatively\, if you wish to clear metadata anyway\, please use \'--force\' option\."
```
Comment (Ostrokrzew) [inline test/functional/api/cas/cli_messages.py:102]: Backslash before comma is not required.

**Review by Ostrokrzew — CHANGES_REQUESTED** [review]

File: test/functional/api/cas/cli_messages.py (hunk 10 above)
Comment (Ostrokrzew) [reply test/functional/api/cas/cli_messages.py:102]: and below

**Review by Ostrokrzew — APPROVED** [review]

File: test/functional/tests/cli/test_zero_metadata_command.py [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+        if output.exit_code != 0:
+            cli_messages.check_stderr_msg(output, cli_messages.no_cas_metadata)
+        else:
+            TestRun.LOGGER.error("Zeroing metadata should fail!")
+
+        file_system = get_device_filesystem_type(core.get_device_id())
+
+        if file_system != filesystem:
+            TestRun.LOGGER.error(f"After zero metadata on core - filesystem ({file_system}) "
+                                 f"not as expected ({filesystem})!")
```
Comment (Deixx) [inline test/functional/tests/cli/test_zero_metadata_command.py:107]: Please reword the message. It could be constructed better ;)

File: test/functional/tests/cli/test_zero_metadata_command.py [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+            TestRun.LOGGER.error("Zeroing metadata should fail!")
+
+
+@pytest.mark.require_disk("cache", DiskTypeSet([DiskType.optane, DiskType.nand]))
+@pytest.mark.require_disk("core", DiskTypeLowerThan("cache"))
+def test_zero_metadata_dirty_data():
+    """
+        title: Test for '--zero-metadata' and dirty data scenario.
+        description: |
+          Test for '--zero-metadata' with&without 'force' option if there are dirty data on cache.
```
Comment (Deixx) [inline test/functional/tests/cli/test_zero_metadata_command.py:126]:
```suggestion
          Test for '--zero-metadata' with & without 'force' option if there are dirty data on cache.
```
Or even `with and without`

File: test/functional/tests/cli/test_zero_metadata_command.py [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+            cli_messages.check_stderr_msg(output, cli_messages.cache_dirty_data)
+        else:
+            TestRun.LOGGER.error("Zeroing metadata without force should fail!")
+
+    with TestRun.step("Zero metadata on cache device with force"):
+        output = TestRun.executor.run(cli.zero_metadata_cmd(str(cache_disk.path), force=True))
+        if output.exit_code == 0:
+            TestRun.LOGGER.info("Zeroing metadata with force ran successfully!")
+        else:
+            TestRun.LOGGER.error("Zero metadata with force should works for cache device!")
```
Comment (Deixx) [inline test/functional/tests/cli/test_zero_metadata_command.py:174]:
```suggestion
            TestRun.LOGGER.error("Zeroing metadata with force should work for cache device!")
```

File: test/functional/tests/cli/test_zero_metadata_command.py [hunk 14]
```diff
@@ -0,0 +1,258 @@
...
+
+        with TestRun.step("Start cache."):
```
Comment (Deixx) [inline test/functional/tests/cli/test_zero_metadata_command.py:176]:
```suggestion
        with TestRun.step("Start cache without 'force' option."):
```

File: test/functional/tests/cli/test_zero_metadata_command.py [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+@pytest.mark.require_disk("core", DiskTypeLowerThan("cache"))
+def test_zero_metadata_dirty_data():
+    """
+        title: Test for '--zero-metadata' and dirty data scenario.
+        description: |
+          Test for '--zero-metadata' with&without 'force' option if there are dirty data on cache.
+        pass_criteria:
+          - Zero metadata without force failed on cache with dirty data.
+          - Zero metadata with force ran successfully on cache with dirty data.
+          - Cache started successfully after operations.
```
Comment (Deixx) [inline test/functional/tests/cli/test_zero_metadata_command.py:130]: Pass criteria should be clear even if separated from each other and the code. Elaborate what operations you have in mind.

File: test/functional/tests/cli/test_zero_metadata_command.py [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+def test_zero_metadata_dirty_shutdown():
+    """
+        title: Test for '--zero-metadata' and dirty shutdown scenario.
+        description: |
+          Test for '--zero-metadata' with and without 'force' option on cache which had been dirty
+          shut down before.
+        pass_criteria:
+          - Zero metadata without force failed on cache after dirty shutdown.
+          - Zero metadata with force ran successfully on cache after dirty shutdown.
+          - Cache started successfully after operations.
```
Comment (Deixx) [inline test/functional/tests/cli/test_zero_metadata_command.py:195]: Same as in previous test.

File: test/functional/tests/cli/test_zero_metadata_command.py [hunk 17]
```diff
@@ -0,0 +1,258 @@
...
+    """
+    with TestRun.step("Prepare cache and core devices."):
+        with TestRun.step("Prepare devices."):
```
Comment (Deixx) [inline test/functional/tests/cli/test_zero_metadata_command.py:198]: Any reason for this step nesting?

File: test/functional/tests/cli/test_zero_metadata_command.py [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+            cli_messages.check_stderr_msg(output, cli_messages.cache_dirty_shutdown)
+        else:
+            TestRun.LOGGER.error("Zeroing metadata without force should fail!")
+
+    with TestRun.step("Zero metadata on cache device with force"):
+        output = TestRun.executor.run(cli.zero_metadata_cmd(str(cache_dev.path), force=True))
+        if output.exit_code == 0:
+            TestRun.LOGGER.info("Zeroing metadata with force ran successfully!")
+        else:
+            TestRun.LOGGER.error("Zero metadata with force should works for cache device!")
```
Comment (Deixx) [inline test/functional/tests/cli/test_zero_metadata_command.py:241]:
```suggestion
            TestRun.LOGGER.error("Zeroing metadata with force should work for cache device!")
```

File: test/functional/tests/cli/test_zero_metadata_command.py [hunk 19]
```diff
@@ -0,0 +1,258 @@
...
+        title: Test for '--zero-metadata' and dirty data scenario.
+        description: |
+          Test for '--zero-metadata' with&without 'force' option if there are dirty data on cache.
+        pass_criteria:
+          - Zero metadata without force failed on cache with dirty data.
+          - Zero metadata with force ran successfully on cache with dirty data.
+          - Cache started successfully after operations.
+    """
+    with TestRun.step("Prepare cache and core devices."):
+        with TestRun.step("Prepare devices."):
```
Comment (Deixx) [inline test/functional/tests/cli/test_zero_metadata_command.py:133]: Any reason for this step nesting?

**Review by Deixx — CHANGES_REQUESTED** [review]

File: test/functional/api/cas/cli_messages.py [hunk 20]
```diff
@@ -83,6 +82,33 @@
...
+    r"Cache instance contains dirty data\. Clearing metadata will result in loss of dirty data\.\n"
+    r"Please load cache instance and flush dirty data in order to preserve them on the core "
+    r"device\.\n"
+    r"Alternatively, if you wish to clear metadata anyway, please use \'--force\' option\."
+]
+
+cache_dirty_shutdown = [
+    r"Cache instance did not shut down cleanly\. It might contain dirty data\. \n"
+    r"Clearing metadata might result in loss of dirty data\. Please recover cache instance\n"
+    r"by loading it and flush dirty data in order to preserve then on the core device\.\n"
```
Comment (jfckm) [inline test/functional/api/cas/cli_messages.py:108]: typo: then -> them

File: test/functional/tests/cli/test_zero_metadata_command.py [hunk 21]
```diff
@@ -0,0 +1,258 @@
...
+          Test for '--zero-metadata' with and without 'force' option on cache which had been dirty
+          shut down before.
+        pass_criteria:
+          - Zero metadata without force failed on cache after dirty shutdown.
+          - Zero metadata with force ran successfully on cache after dirty shutdown.
+          - Cache started successfully after operations.
+    """
+    with TestRun.step("Prepare cache and core devices."):
+        with TestRun.step("Prepare devices."):
+            cache_disk = TestRun.disks['cache']
```
Comment (jfckm) [inline test/functional/tests/cli/test_zero_metadata_command.py:199]: You could use `prepare_devices()` function here and just make it return the cache_disk for unplugging, you could just ignore this in other tests

File: test/functional/tests/cli/test_zero_metadata_command.py [hunk 22]
```diff
@@ -0,0 +1,258 @@
...
+        try:
+            cache = casadm.start_cache(cache_dev, CacheMode.WT)
+            TestRun.LOGGER.info("Cache started successfully.")
+        except CmdException:
+            TestRun.LOGGER.error("Start cache failed.")
+
+
+def prepare_devices():
+    cache_disk = TestRun.disks['cache']
+    cache_disk.create_partitions([Size(2, Unit.GibiByte)])
```
Comment (jfckm) [inline test/functional/tests/cli/test_zero_metadata_command.py:253]: Given the nature of the tests I think you could decrease the cache size to maybe 100MiB - that would save some execution time

File: test/functional/api/cas/cli_messages.py (hunk 20 above)
Comment (karolinavelkaja) [reply test/functional/api/cas/cli_messages.py:108]: nice catch! The typo is present in casadm as well so fix is needed in both places

File: test/functional/tests/cli/test_zero_metadata_command.py (hunk 1 above)
Comment (Deixx) [reply test/functional/tests/cli/test_zero_metadata_command.py]: Instead of using cli.zero_metadata_cmd() use casadm.zero_metadata() and try..except approach

File: test/functional/api/cas/cli_messages.py (hunk 20 above)
Comment (Deixx) [reply test/functional/api/cas/cli_messages.py:108]: @karolinavelkaja please create a low priority bug for the typo in casadm

File: test/functional/tests/cli/test_zero_metadata_command.py (hunk 19 above)
Comment (karolinavelkaja) [reply test/functional/tests/cli/test_zero_metadata_command.py:133]: omission, removing!:)

File: test/functional/api/cas/cli_messages.py (hunk 20 above)
Comment (karolinavelkaja) [reply test/functional/api/cas/cli_messages.py]: Test consists fix for typo, buf created and should be linked: https://github.com/Open-CAS/open-cas-linux/issues/917

Comment (karolinavelkaja) [reply test/functional/api/cas/cli_messages.py]: Fix in casadm merged, test is up-to date (ready to review&merge)

File: test/functional/tests/cli/test_zero_metadata_command.py [hunk 23]
```diff
@@ -0,0 +1,256 @@
...
+        cache_dev, core_disk, cache_disk = prepare_devices()
+
+    with TestRun.step("Create filesystem on core device."):
+        core_disk.create_filesystem(filesystem)
+
+    with TestRun.step("Start cache and add core."):
+        cache = casadm.start_cache(cache_dev, force=True)
+        core = cache.add_core(core_disk)
+
+    with TestRun.step("Zeroing metadata on core device an validating filesystem"):
```
Comment (robertbaldyga) [inline test/functional/tests/cli/test_zero_metadata_command.py:98]: s/an/and

**Review by robertbaldyga — CHANGES_REQUESTED** [review]
I don't see test where we try to zero metadata with force on cache device while cache is still running.

File: test/functional/tests/cli/test_zero_metadata_command.py [hunk 24]
```diff
@@ -0,0 +1,256 @@
...
+
+    with TestRun.step("Start cache."):
+        cache = casadm.start_cache(cache_dev, force=True)
+
+    with TestRun.step("Try to zero metadata and validate error message."):
+        try:
+            casadm.zero_metadata(cache_dev)
+            TestRun.LOGGER.error("Zeroing metadata should fail!")
+        except CmdException as e:
+            cli_messages.check_stderr_msg(e.output, cli_messages.unavailable_device)
```
Comment (Deixx) [inline test/functional/tests/cli/test_zero_metadata_command.py:45]: add a step where you try to zero metadata with force

**Review by Deixx — CHANGES_REQUESTED** [review]

**Review by robertbaldyga — APPROVED** [review]
